include/wil/com_apartment_variable.h

```diff
@@ -0,0 +1,377 @@
+    {
+        return std::any_cast<std::invoke_result_t<F>&>(details::get_for_current_com_apartment(std::addressof(createFn), [createFn]()
+        {
+            return createFn();
+        }));
+    }
+
+    // Remove the variable including its entry in the map. This has no
+    // effect if the variable was not stored based on a call to get_for_current_com_apartment
+    // from the current apartment.
+    template<typename F>
+    void reset_for_current_com_apartment(F const* createFn)
+    {
+        details::reset_for_current_com_apartment(std::addressof(createFn));
+    }
+
+    // Replace a variable with a new value. This requires (ensured via fail fast) that the
+    // variable is already stored based on a call to get_for_current_com_apartment in the
+    // current apartment.
+    template<typename F>
+    void reset_for_current_com_apartment(F const* createFn, std::invoke_result_t<F>&& newValue)
+    {
+        details::reset_for_current_com_apartment(std::addressof(createFn), std::forward<std::invoke_result_t<F>>(newValue));
+    }
+
+    // for testing
+    size_t apartment_count() { return details::s_apartmentStorage.size(); }
+    size_t apartment_variable_count()
+    {
+        const auto apartmentId = details::GetApartmentId();
+
+        auto lock = winrt::slim_lock_guard(details::s_lock);
+        auto storage = details::s_apartmentStorage.find(apartmentId);
+        if (storage != details::s_apartmentStorage.end())
+        {
+            return storage->second.size();
+        }
+        return 0;
+    }
+
+    namespace details
+    {
+        struct apt_var_platform
+        {
+            static unsigned long long GetApartmentId()
+            {
+                unsigned long long apartmentId{};
+                FAIL_FAST_IF_FAILED(RoGetApartmentIdentifier(&apartmentId));
+                return apartmentId;
+            }
+
+            static void RegisterForApartmentShutdown(IApartmentShutdown* observer)
+            {
+                unsigned long long id{};
+                APARTMENT_SHUTDOWN_REGISTRATION_COOKIE cookie{}; // leaked, COM run down releases all instances
+                THROW_IF_FAILED(RoRegisterForApartmentShutdown(observer, &id, &cookie));
+            }
+        };
+
+        template<typename test_hook = apt_var_platform>
+        struct apartment_variable_base
+        {
+            inline static winrt::slim_mutex s_lock;
+            // Apartment id -> variable storage.
+            // Variables are stored using the address of the global variable as the key.
+            inline static std::unordered_map<unsigned long long,
+                std::unordered_map<wil::details::apartment_variable_base<test_hook>*, std::any>>
+                s_apartmentStorage;
+
+            apartment_variable_base()
+            {
+                // TODO: ensure this is a global, investigate cases where address is not reliable due to re-use as stack variables.

```

Comment:
If the destructor cleans up all copies, then it's okay for this to be a stack or heap variable.

Requiring this to be a global means that you can't have per-instance apartment variables, which would also be useful.